client/components/infinite-scroll/index.js

```diff
@@ -3,53 +3,92 @@
 /**
  * External dependencies
  */
-import { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
-import { throttle } from 'lodash';
+import { throttle, defer } from 'lodash';
+import afterLayoutFlush from 'lib/after-layout-flush';
 
 const SCROLL_CHECK_RATE_IN_MS = 400;
 
-export default class InfiniteScroll extends Component {
+export default class InfiniteScroll extends React.Component {
 	static propTypes = {
 		nextPageMethod: PropTypes.func.isRequired,
 	};
 
-	checkScrollPositionHandler = throttle(
-		() => this.checkScrollPosition( true ),
-		SCROLL_CHECK_RATE_IN_MS
-	);
+	observedElement = React.createRef();
 
 	componentDidMount() {
-		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
-		this.checkScrollPosition( false );
+		this.hasIntersectionObserver = 'IntersectionObserver' in window;
+
+		if ( this.hasIntersectionObserver && this.observedElement.current ) {
+			this.observer = new IntersectionObserver( this.handleIntersection, {
+				rootMargin: '100%',
+				threshold: 1.0,
+			} );
+			this.observer.observe( this.observedElement.current );
+		} else {
+			window.addEventListener( 'scroll', this.checkScrollPositionHandler );
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
+		}
 	}
 
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		if ( ! this.hasIntersectionObserver && ! this.deferredTimer ) {
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
+		}
 	}
 
 	componentWillUnmount() {
+		// Clean up scroll codepath.
 		window.removeEventListener( 'scroll', this.checkScrollPositionHandler );
+		this.clearTimers();
+
+		// Clean up IntersectionObserver codepath.
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
 	}
 
+	handleIntersection = entries => {
+		if ( entries && entries[ 0 ] && entries[ 0 ].isIntersecting ) {
+			const { boundingClientRect, rootBounds } = entries[ 0 ];
+			const triggeredByScroll = boundingClientRect.bottom >= rootBounds.bottom;
```

Comment:
When I scroll down the `/themes` page (logged-out version), the `triggeredByScroll` param is always `false`, although the fetch was triggered by scroll.

We use the flag to send analytics about user behavior on the page.

Reply:
I experimented with intersection observers a little and discovered that the callback is not called immediately and synchronously at the very moment the element scrolls into view. Therefore the exact equality check
```js
boundingClientRect.bottom === rootBounds.bottom
```
is not reliable at all. The element can be scrolled a few more pixels into the viewport at the time when the intersection observer reports the intersection.

The following condition:
```js
isIntersecting && boundingClientRect.bottom > rootBounds.bottom
```
can never be true for a zero-height element and/or threshold of 1.0. If the bottom edge of the observed element is below the root bottom edge, then it cannot be intersecting.

Reply:
Yes, this logic was completely wrong, thank you for looking so deeply into this. I should have tested this more thoroughly.

I've completely reworked the logic here, by instead keeping track of when the anchor element first stops intersecting with the viewport. I think it's safe to assume that any fetches after that are triggered by scroll.